twisted/internet/interfaces.py

```diff
@@ -2207,6 +2207,28 @@ def clientConnectionForTLS(tlsProtocol):
 
 
 
+class IProtocolNegotiationFactory(Interface):
```

Comment:
> Shouldn't something tell you?

Yes. Specifically, `INegotiated.negotiatedProtocol`. That already exists. =)

This PR is essentially a reworking of the previous NPN/ALPN change, which already made it possible to perform NPN/ALPN negotiation. We're just extending it to allow somewhat-automated protocol offering by factories.